bibliopixel/drivers/curses.py

```diff
@@ -0,0 +1,91 @@
+import curses as _curses
+import itertools, queue, weakref
+from . driver_base import DriverBase
+
+
+class Curses(DriverBase):
+    DRIVERS = weakref.WeakSet()
```

Comment:
What is this?

Reply:
Ah, this is a very common Python idiom.  I should have used it before.

I need to make sure that there is at most one Curses driver - and I need to be able to retrieve that driver if it exists so I can use it in "main".

The second condition means I need to store that driver in a global variable.  But if I put it in a regular variable or a list, then it will stay around after I'm finished with it...

Instead I go to [weak references](https://docs.python.org/3/library/weakref.html).  Basically, weak references do _not_ keep an object alive.  Another way to think of it is that elements are automatically deleted from the `WeakSet` when all other references to it go away.

I could have done it with a single `weakref.ref`, and not the set, but `WeakSet` has a particularly easy API and I was "lazy" - also, the code is clearer.

We do exactly the same thing in project.py.  There, the set makes more sense because one day we might even be able to run more than one project...

Reply:
OK... cool. That was new to me.